README.md
# IS31FL3193 RGB LED driver

I2C driver for the IS31FL3193 RGB LED driver

## [Documentation](https://docs.rs/IS31FL3193)

## [Changelog](CHANGELOG.md)

## [Examples](examples)

This crate uses [`probe-run`](https://crates.io/crates/probe-run) to run the examples. Once set up,
it should be as simple as `cargo run --example <example name> --release`.

From [`examples/rppico_blink.rs`](examples/rppico_blink.rs):

```rust
#![no_std]
#![no_main]
use fugit::RateExtU32;
use hal::pac;

use defmt_rtt as _;
use panic_probe as _;

use is31fl3193::*;
use rp2040_hal as hal;
use rp_pico::{
    entry,
    hal::{Clock, I2C},
    XOSC_CRYSTAL_FREQ,
};

#[entry]
fn main() -> ! {
    let pac = pac::Peripherals::take().unwrap();
    // Set up the watchdog driver - needed by the clock setup code
    let mut watchdog = hal::Watchdog::new(pac.WATCHDOG);
    // Soft-reset does not release the hardware spinlocks
    // Release them now to avoid a deadlock after debug or watchdog reset
    unsafe {
        rp_pico::hal::sio::spinlock_reset();
    }
    let mut resets = pac.RESETS;

    // Configure the clocks
    let clocks = hal::clocks::init_clocks_and_plls(
        XOSC_CRYSTAL_FREQ,
        pac.XOSC,
        pac.CLOCKS,
        pac.PLL_SYS,
        pac.PLL_USB,
        &mut resets,
        &mut watchdog,
    )
    .ok()
    .unwrap();

    let sio = hal::Sio::new(pac.SIO);
    let pins = rp_pico::Pins::new(pac.IO_BANK0, pac.PADS_BANK0, sio.gpio_bank0, &mut resets);
    // ----------------------------------------------------------------
    // ----I2C0---- configuration
    // ----------------------------------------------------------------
    let i2c0 = I2C::i2c0(
        pac.I2C0,
        pins.gpio20.into_mode(),
        pins.gpio21.into_mode(),
        400.kHz(),
        &mut resets,
        clocks.peripheral_clock.freq(),
    );

    let mut led_drv = IS31FL3193::new(i2c0, ADPin::ToGND);

    //
    // 1, Example PWM
    //
    // Before exiting from shutdown, set the maximum current allowed
    led_drv.set_max_current(Intensity::MA5).unwrap();
    // Set the mode we want (PWM is default)
    led_drv.set_mode(Mode::PWM).unwrap();
    // Set PWM values
    led_drv.set_pwm(100, 0, 75).unwrap();
    // Enable LEDS and go out of shutdown mode
    led_drv.shutdown(true, false).unwrap();

    // //
    // // 2, Example Breath
    // //
    // // Before exiting from shutdown, set the maximum current allowed
    // led_drv.set_max_current(Intensity::MA5).unwrap();
    // // PWM setting is also necessary for breathing mode
    // led_drv.set_pwm(50, 50, 50).unwrap();
    // // Set the mode we want (PWM is default)
    // led_drv
    //     .set_mode(Mode::Breath(BreathingMode::Auto, Marking::Off))
    //     .unwrap();
    // // Enable LEDS and go out of shutdown mode
    // led_drv.shutdown(true, false).unwrap();

    // //
    // // 3, Example Breath with differents timing for each color
    // //
    // // Before exiting from shutdown, set the maximum current allowed
    // led_drv.set_max_current(Intensity::MA5).unwrap();
    // // Set custom timings
    // led_drv
    //     .set_timing(
    //         Channel::Led1,
    //         T0::MS260,
    //         T1::MS130,
    //         T2::MS130,
    //         T3::MS260,
    //         T4::MS130,
    //     )
    //     .unwrap();
    // led_drv
    //     .set_timing(
    //         Channel::Led2,
    //         T0::MS130,
    //         T1::MS130,
    //         T2::MS130,
    //         T3::MS520,
    //         T4::MS130,
    //     )
    //     .unwrap();
    // led_drv
    //     .set_timing(
    //         Channel::Led3,
    //         T0::MS130,
    //         T1::MS1040,
    //         T2::MS130,
    //         T3::MS130,
    //         T4::MS130,
    //     )
    //     .unwrap();
    // // PWM setting is also necessary for breathing mode
    // led_drv.set_pwm(50, 50, 50).unwrap();
    // // Set the mode we want (PWM is default)
    // led_drv
    //     .set_mode(Mode::Breath(BreathingMode::Auto, Marking::Off))
    //     .unwrap();
    // // Enable LEDS and go out of shutdown mode
    // led_drv.shutdown(true, false).unwrap();

    defmt::info!("Done, entering loop");
    loop {}
}

// same panicking *behavior* as `panic-probe` but doesn't print a panic message
// this prevents the panic message being printed *twice* when `defmt::panic` is invoked
#[defmt::panic_handler]
fn panic() -> ! {
    cortex_m::asm::udf()
}

/// Terminates the application and makes `probe-run` exit with exit-code = 0
pub fn exit() -> ! {
    loop {
        cortex_m::asm::bkpt();
    }
}
```

## License

Licensed under either of

- Apache License, Version 2.0 ([LICENSE-APACHE](LICENSE-APACHE) or
  http://www.apache.org/licenses/LICENSE-2.0)
- MIT license ([LICENSE-MIT](LICENSE-MIT) or http://opensource.org/licenses/MIT)

at your option.

### Contribution

Unless you explicitly state otherwise, any contribution intentionally submitted for inclusion in the
work by you, as defined in the Apache-2.0 license, shall be dual licensed as above, without any
additional terms or conditions.


CHANGELOG.md
# Changelog

[`is31fl3193`](https://crates.io/crates/is31fl3193) is a no-std driver written in Rust for the is31fl3193 IC RGB LED driver.

<!-- next-header -->

## [Unreleased] - ReleaseDate

## 0.1.1

### Added

- Documentation

## 0.1.0

### Added

- Initial release

### Fixed

### Changed



src/lib.rs
//! IS31FL3193 RGB LED driver
//!
//! This crate provides a driver interface for the IS31FL3193 IC from LUMISSIL
//!
//! # Examples
//!
//! An example of implementation for the Raspberry Pi Pico can be found in examples folder

#![no_std]
mod command;
use command::Command;
use embedded_hal as hal;

/// All possible errors in this crate
#[derive(Clone, Debug)]
pub enum IS31FL3193Error<E> {
    /// I2C bus error
    I2c(E),
}

/// Hardware AD pin (pin 7) connection
pub enum ADPin {
    ToGND,
    ToVCC,
    ToSCL,
    ToSDA,
}

/// is31fl3193 driver.
#[derive(Copy, Clone, Debug)]
pub struct IS31FL3193<I2C> {
    i2c: I2C,
    address: u8,
}
impl<I2C, E> IS31FL3193<I2C>
where
    I2C: hal::blocking::i2c::Write<Error = E>,
{
    /// Create a new I2C interface with an address that
    /// corresponds to the AD pin hardware connection
    pub fn new(i2c: I2C, adpin: ADPin) -> IS31FL3193<I2C> {
        let address = match adpin {
            ADPin::ToGND => 0xD0 >> 1,
            ADPin::ToVCC => 0xD6 >> 1,
            ADPin::ToSCL => 0xD2 >> 1,
            ADPin::ToSDA => 0xD4 >> 1,
        };
        Self { i2c, address }
    }
    /// Consume the interface and return the underlying peripherial driver
    pub fn release(self) -> I2C {
        self.i2c
    }
    /// Set the maximum current for all the three outputs
    pub fn set_max_current(&mut self, intensity: Intensity) -> Result<(), IS31FL3193Error<E>> {
        self.send(Command::Current(intensity))
    }
    /// Set PWM for each of the three output
    pub fn set_pwm(&mut self, led1: u8, led2: u8, led3: u8) -> Result<(), IS31FL3193Error<E>> {
        self.send(Command::PWMLed1(led1))?;
        self.send(Command::PWMLed2(led2))?;
        self.send(Command::PWMLed3(led3))?;
        self.send(Command::PWMUpdate)
    }
    /// Set PWM for LED1 (OUT1)
    pub fn set_pwm_led1(&mut self, led1: u8) -> Result<(), IS31FL3193Error<E>> {
        self.send(Command::PWMLed1(led1))?;
        self.send(Command::PWMUpdate)
    }
    /// Set PWM for LED2 (OUT2)
    pub fn set_pwm_led2(&mut self, led2: u8) -> Result<(), IS31FL3193Error<E>> {
        self.send(Command::PWMLed1(led2))?;
        self.send(Command::PWMUpdate)
    }
    /// Set PWM for LED3 (OUT3)
    pub fn set_pwm_led3(&mut self, led3: u8) -> Result<(), IS31FL3193Error<E>> {
        self.send(Command::PWMLed1(led3))?;
        self.send(Command::PWMUpdate)
    }
    /// Set the breath timing for a given output
    pub fn set_timing(
        &mut self,
        led: Channel,
        t0: T0,
        t1: T1,
        t2: T2,
        t3: T3,
        t4: T4,
    ) -> Result<(), IS31FL3193Error<E>> {
        match led {
            Channel::Led1 => {
                self.send(Command::T0Led1(t0))?;
                self.send(Command::T1T2Led1(t1, t2))?;
                self.send(Command::T3T4Led1(t3, t4))?;
            }
            Channel::Led2 => {
                self.send(Command::T0Led2(t0))?;
                self.send(Command::T1T2Led2(t1, t2))?;
                self.send(Command::T3T4Led2(t3, t4))?;
            }
            Channel::Led3 => {
                self.send(Command::T0Led3(t0))?;
                self.send(Command::T1T2Led3(t1, t2))?;
                self.send(Command::T3T4Led3(t3, t4))?;
            }
        }
        self.send(Command::BreathTimingUpdate)
    }
    /// Set the mode. There are two modes, PWM and breathing (see DS)
    pub fn set_mode(&mut self, mode: Mode) -> Result<(), IS31FL3193Error<E>> {
        match mode {
            Mode::PWM => self.send(Command::LedMode(false)),
            Mode::Breath(breath_mode, marking) => {
                self.send(Command::LedMode(true))?;
                match breath_mode {
                    BreathingMode::Auto => match marking {
                        Marking::Off => {
                            self.send(Command::Breathing(false, false, false, Channel::Led1))
                        }
                        Marking::On(channel) => {
                            self.send(Command::Breathing(false, false, true, channel))
                        }
                    },
                    BreathingMode::OneCycle => match marking {
                        Marking::Off => {
                            self.send(Command::Breathing(true, true, false, Channel::Led1))
                        }
                        Marking::On(channel) => {
                            self.send(Command::Breathing(true, true, true, channel))
                        }
                    },
                    BreathingMode::RampToOn => match marking {
                        Marking::Off => {
                            self.send(Command::Breathing(true, false, false, Channel::Led1))
                        }
                        Marking::On(channel) => {
                            self.send(Command::Breathing(true, false, true, channel))
                        }
                    },
                }
            }
        }
    }
    /// Enable/disable all LEDS and set/clear shutdown
    pub fn shutdown(
        &mut self,
        enable_all_leds: bool,
        shutdown: bool,
    ) -> Result<(), IS31FL3193Error<E>> {
        self.send(Command::ShutDown(enable_all_leds, shutdown))
    }
    fn send(&mut self, cmd: Command) -> Result<(), IS31FL3193Error<E>> {
        self.i2c
            .write(self.address, &cmd.as_bytes())
            .map_err(IS31FL3193Error::I2c)
    }
}

/// Maximum currents
#[repr(u8)]
#[derive(Copy, Clone)]
pub enum Intensity {
    MA5 = 2 << 2,
    MA10 = 1 << 2,
    MA17 = 4 << 2,
    MA30 = 3 << 2,
    MA42 = 0 << 2,
}
/// IC modes
#[allow(clippy::upper_case_acronyms)]
#[derive(Copy, Clone)]
pub enum Mode {
    PWM,
    Breath(BreathingMode, Marking),
}
/// Breathing. Auto for continuous variations, OneCycle for one shot variation,
/// RampToOn for ramping from 0 to PWM value
#[repr(u8)]
#[derive(Copy, Clone)]
pub enum BreathingMode {
    Auto,
    OneCycle,
    RampToOn,
}
/// Enable/disable hardware output marking for a given output
#[derive(Copy, Clone)]
pub enum Marking {
    Off,
    On(Channel),
}
/// Output selection
#[repr(u8)]
#[derive(Copy, Clone)]
pub enum Channel {
    Led1 = 0,
    Led2 = 1,
    Led3 = 2,
}
/// Timing. T0 is an offset time period which runs only once at the start of the cycle
#[repr(u8)]
#[derive(Copy, Clone)]
pub enum T0 {
    MS0 = 0,
    MS130 = 1,
    MS260 = 2,
    MS520 = 3,
    MS1040 = 4,
    MS2080 = 5,
    MS4160 = 6,
    MS8320 = 7,
    MS16640 = 8,
    MS33280 = 9,
    MS66560 = 10,
}
/// Timing. T1 is the rising time
#[repr(u8)]
#[derive(Copy, Clone)]
pub enum T1 {
    MS130 = 0 << 5,
    MS260 = 1 << 5,
    MS520 = 2 << 5,
    MS1040 = 3 << 5,
    MS2080 = 4 << 5,
    MS4160 = 5 << 5,
    MS8320 = 6 << 5,
    MS16640 = 7 << 5,
}
/// Timing. T2 is the steady ON time
#[repr(u8)]
#[derive(Copy, Clone)]
pub enum T2 {
    MS0 = 0 << 1,
    MS130 = 1 << 1,
    MS260 = 2 << 1,
    MS520 = 3 << 1,
    MS1040 = 4 << 1,
    MS2080 = 5 << 1,
    MS4160 = 6 << 1,
    MS8320 = 7 << 1,
    MS16640 = 8 << 1,
}
/// Timing. T3 is the falling time
#[repr(u8)]
#[derive(Copy, Clone)]
pub enum T3 {
    MS130 = 0,
    MS260 = 1 << 5,
    MS520 = 2 << 5,
    MS1040 = 3 << 5,
    MS2080 = 4 << 5,
    MS4160 = 5 << 5,
    MS8320 = 6 << 5,
    MS16640 = 7 << 5,
}
/// Timing. T4 is the steady OFF time
#[repr(u8)]
#[derive(Copy, Clone)]
pub enum T4 {
    MS0 = 0 << 1,
    MS130 = 1 << 1,
    MS260 = 2 << 1,
    MS520 = 3 << 1,
    MS1040 = 4 << 1,
    MS2080 = 5 << 1,
    MS4160 = 6 << 1,
    MS8320 = 7 << 1,
    MS16640 = 8 << 1,
    MS33280 = 9 << 1,
    MS66560 = 10 << 1,
}


src/command.rs
//! IS31FL3193 commands

use crate::{Channel, Intensity, T0, T1, T2, T3, T4};

// Shamefully taken from https://github.com/jamwaffles/ssd1306

/// Commands
#[derive(Copy, Clone)]
#[allow(dead_code)]
pub(crate) enum Command {
    /// First bool: All channels enable/disable
    /// Second bool: false: normal operation, true: software shutdown mode
    ShutDown(bool, bool),
    /// Breathing control
    /// First bool: Ramping mode enable
    /// Second bool: Hold time, false:T2, true:T4
    /// Third bool: Breathing mark enable
    /// Channel: Channel selection
    Breathing(bool, bool, bool, Channel),
    /// RGB mode, PWM(false) or breathing(true)
    LedMode(bool),
    /// LED Maximum current setting
    Current(Intensity),
    /// PWM settings
    PWMLed1(u8),
    PWMLed2(u8),
    PWMLed3(u8),
    /// Update PWMs settings to Leds
    PWMUpdate,
    /// T0:Time for breathing
    T0Led1(T0),
    T0Led2(T0),
    T0Led3(T0),
    /// T1,T2: Time for breathing
    T1T2Led1(T1, T2),
    T1T2Led2(T1, T2),
    T1T2Led3(T1, T2),
    /// T3,T4: Time for breathing
    T3T4Led1(T3, T4),
    T3T4Led2(T3, T4),
    T3T4Led3(T3, T4),
    /// Update T0,T1,T2,T3,T4 breathing times to LEDs
    BreathTimingUpdate,
    /// Power on/off LEDS
    /// /// First bool: LED1, second: LED2, Third: LED3
    Power(bool, bool, bool),
    /// Reset to default (power on default)
    Reset,
}

impl Command {
    /// Send command to is31fl3193
    pub(crate) fn as_bytes(self) -> [u8; 2] {
        // Transform command into a fixed size array of 7 u8 and the real length for sending
        use Command::*;
        match self {
            ShutDown(en, ssd) => [0x00, (en as u8) << 5 | ssd as u8],
            Breathing(rm, ht, bme, css) => [
                0x01,
                (rm as u8) << 5 | (ht as u8) << 4 | (bme as u8) << 2 | (css as u8),
            ],
            LedMode(breath) => [0x02, (breath as u8) << 5],
            Current(intensity) => [0x03, intensity as u8],
            PWMLed1(pwm) => [0x04, pwm],
            PWMLed2(pwm) => [0x05, pwm],
            PWMLed3(pwm) => [0x06, pwm],
            PWMUpdate => [0x07, 0],
            T0Led1(t0) => [0x0A, (t0 as u8) << 4],
            T0Led2(t0) => [0x0B, (t0 as u8) << 4],
            T0Led3(t0) => [0x0C, (t0 as u8) << 4],
            T1T2Led1(t1, t2) => [0x10, (t1 as u8) | t2 as u8],
            T1T2Led2(t1, t2) => [0x11, (t1 as u8) | t2 as u8],
            T1T2Led3(t1, t2) => [0x12, (t1 as u8) | t2 as u8],
            T3T4Led1(t3, t4) => [0x16, (t3 as u8) | t4 as u8],
            T3T4Led2(t3, t4) => [0x17, (t3 as u8) | t4 as u8],
            T3T4Led3(t3, t4) => [0x18, (t3 as u8) | t4 as u8],
            BreathTimingUpdate => [0x1C, 0],
            Power(led1, led2, led3) => [0x1D, (led1 as u8) | (led2 as u8) << 1 | (led3 as u8) << 2],
            Reset => [0x2F, 0],
        }
    }
}


examples/rppico_blink.rs
#![no_std]
#![no_main]
use fugit::RateExtU32;
use hal::pac;

use defmt_rtt as _;
use panic_probe as _;

use is31fl3193::*;
use rp2040_hal as hal;
use rp_pico::{
    entry,
    hal::{Clock, I2C},
    XOSC_CRYSTAL_FREQ,
};

#[entry]
fn main() -> ! {
    let pac = pac::Peripherals::take().unwrap();
    // Set up the watchdog driver - needed by the clock setup code
    let mut watchdog = hal::Watchdog::new(pac.WATCHDOG);
    // Soft-reset does not release the hardware spinlocks
    // Release them now to avoid a deadlock after debug or watchdog reset
    unsafe {
        rp_pico::hal::sio::spinlock_reset();
    }
    let mut resets = pac.RESETS;

    // Configure the clocks
    let clocks = hal::clocks::init_clocks_and_plls(
        XOSC_CRYSTAL_FREQ,
        pac.XOSC,
        pac.CLOCKS,
        pac.PLL_SYS,
        pac.PLL_USB,
        &mut resets,
        &mut watchdog,
    )
    .ok()
    .unwrap();

    let sio = hal::Sio::new(pac.SIO);
    let pins = rp_pico::Pins::new(pac.IO_BANK0, pac.PADS_BANK0, sio.gpio_bank0, &mut resets);
    // ----------------------------------------------------------------
    // ----I2C0---- configuration
    // ----------------------------------------------------------------
    let i2c0 = I2C::i2c0(
        pac.I2C0,
        pins.gpio20.into_mode(),
        pins.gpio21.into_mode(),
        400.kHz(),
        &mut resets,
        clocks.peripheral_clock.freq(),
    );

    let mut led_drv = IS31FL3193::new(i2c0, ADPin::ToGND);

    //
    // 1, Example PWM
    //
    // Before exiting from shutdown, set the maximum current allowed
    led_drv.set_max_current(Intensity::MA5).unwrap();
    // Set the mode we want (PWM is default)
    led_drv.set_mode(Mode::PWM).unwrap();
    // Set PWM values
    led_drv.set_pwm(100, 0, 75).unwrap();
    // Enable LEDS and go out of shutdown mode
    led_drv.shutdown(true, false).unwrap();

    // //
    // // 2, Example Breath
    // //
    // // Before exiting from shutdown, set the maximum current allowed
    // led_drv.set_max_current(Intensity::MA5).unwrap();
    // // PWM setting is also necessary for breathing mode
    // led_drv.set_pwm(50, 50, 50).unwrap();
    // // Set the mode we want (PWM is default)
    // led_drv
    //     .set_mode(Mode::Breath(BreathingMode::Auto, Marking::Off))
    //     .unwrap();
    // // Enable LEDS and go out of shutdown mode
    // led_drv.shutdown(true, false).unwrap();

    // //
    // // 3, Example Breath with differents timing for each color
    // //
    // // Before exiting from shutdown, set the maximum current allowed
    // led_drv.set_max_current(Intensity::MA5).unwrap();
    // // Set custom timings
    // led_drv
    //     .set_timing(
    //         Channel::Led1,
    //         T0::MS260,
    //         T1::MS130,
    //         T2::MS130,
    //         T3::MS260,
    //         T4::MS130,
    //     )
    //     .unwrap();
    // led_drv
    //     .set_timing(
    //         Channel::Led2,
    //         T0::MS130,
    //         T1::MS130,
    //         T2::MS130,
    //         T3::MS520,
    //         T4::MS130,
    //     )
    //     .unwrap();
    // led_drv
    //     .set_timing(
    //         Channel::Led3,
    //         T0::MS130,
    //         T1::MS1040,
    //         T2::MS130,
    //         T3::MS130,
    //         T4::MS130,
    //     )
    //     .unwrap();
    // // PWM setting is also necessary for breathing mode
    // led_drv.set_pwm(50, 50, 50).unwrap();
    // // Set the mode we want (PWM is default)
    // led_drv
    //     .set_mode(Mode::Breath(BreathingMode::Auto, Marking::Off))
    //     .unwrap();
    // // Enable LEDS and go out of shutdown mode
    // led_drv.shutdown(true, false).unwrap();

    defmt::info!("Done, entering loop");
    loop {}
}

// same panicking *behavior* as `panic-probe` but doesn't print a panic message
// this prevents the panic message being printed *twice* when `defmt::panic` is invoked
#[defmt::panic_handler]
fn panic() -> ! {
    cortex_m::asm::udf()
}

/// Terminates the application and makes `probe-run` exit with exit-code = 0
pub fn exit() -> ! {
    loop {
        cortex_m::asm::bkpt();
    }
}
